File: test_endpoints.py
import requests
import json
import time

BASE_URL = "http://localhost:2020/resourcivo"
API_DOCS_URL = f"{BASE_URL}/api-docs"

def register_and_login():
    # Register
    register_url = f"{BASE_URL}/api/auth/register"
    user_data = {
        "username": "test_user_" + str(int(time.time())),
        "email": f"test_{int(time.time())}@example.com",
        "password": "password123",
        "role": "STUDENT"
    }
    
    print(f"Registering user: {user_data['username']}")
    try:
        resp = requests.post(register_url, json=user_data)
        print(f"Registration Response: {resp.status_code} - {resp.text}")
    except Exception as e:
        print(f"Registration error: {e}")

    # Login
    login_url = f"{BASE_URL}/api/auth/login"
    login_data = {
        "username": user_data["username"],
        "password": user_data["password"]
    }
    
    print("Logging in...")
    resp = requests.post(login_url, json=login_data)
    if resp.status_code == 200:
        token = resp.json().get("token")
        print("Login successful, token received.")
        return token
    else:
        print(f"Login failed: {resp.text}")
        return None

def test_endpoints(token):
    print(f"Fetching API docs from {API_DOCS_URL}...")
    try:
        resp = requests.get(API_DOCS_URL)
        if resp.status_code != 200:
            print(f"Failed to fetch API docs: {resp.status_code}")
            return
        
        api_docs = resp.json()
        paths = api_docs.get("paths", {})
        
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json"
        }
        
        results = []
        
        print(f"Found {len(paths)} paths. Testing GET endpoints...")
        
        for path, methods in paths.items():
            if "get" in methods:
                # Handle path parameters
                test_path = path.replace("{id}", "1").replace("{userId}", "1").replace("{studentId}", "1")
                
                # Skip other complex parameters for now
                if "{" in test_path:
                    print(f"Skipping complex path: {path}")
                    continue
                
                url = f"{BASE_URL}{test_path}"
                try:
                    r = requests.get(url, headers=headers)
                    status = r.status_code
                    # 200 OK, 404 Not Found (if ID 1 doesn't exist), 403 Forbidden (role mismatch) are all "working" in the sense that the endpoint is reachable.
                    # 500 is a failure.
                    result = "PASS" if status < 500 else "FAIL"
                    print(f"[{result}] {status} - {path}")
                    results.append((path, status, result))
                except Exception as e:
                    print(f"[FAIL] Error - {path}: {e}")
                    results.append((path, "ERR", "FAIL"))
                    
        print("\nSummary:")
        pass_count = sum(1 for _, _, r in results if r == "PASS")
        print(f"Total GET Endpoints Tested: {len(results)}")
        print(f"Passed (reachable): {pass_count}")
        print(f"Failed (500/Error): {len(results) - pass_count}")
        
    except Exception as e:
        print(f"Error testing endpoints: {e}")

if __name__ == "__main__":
    token = register_and_login()
    if token:
        test_endpoints(token)
    else:
        print("Cannot proceed without token.")

File: test-endpoints.js
async function testEndpoints(baseUrl) {
    console.log(`\n=== Testing Endpoints on ${baseUrl} ===`);
    let token;

    // 1. Login to get token
    try {
        const loginRes = await fetch(`${baseUrl}/api/auth/login`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Origin': 'http://localhost:3003'
            },
            body: JSON.stringify({ username: 'username', password: 'password' })
        });

        if (!loginRes.ok) {
            console.error(`[FAIL] Login Failed: ${loginRes.status}`);
            return;
        }
        const data = await loginRes.json();
        token = data.token;
        console.log('[PASS] Login Success');
    } catch (e) {
        console.error(`[FAIL] Network Error during Login: ${e.message}`);
        return;
    }

    const headers = {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json'
    };

    // 2. Test Student Search (Directory)
    // Formerly /api/student/search (SecurityConfig had /api/students/**)
    console.log('\nTesting Student Search (/api/student/search)...');
    try {
        const res = await fetch(`${baseUrl}/api/student/search`, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify({}) // Empty filter
        });
        if (res.ok) console.log(`[PASS] Student Search: ${res.status}`);
        else console.error(`[FAIL] Student Search: ${res.status} - ${await res.text()}`);
    } catch (e) { console.error(`[FAIL] Student Search Error: ${e.message}`); }

    // 3. Test Library Books (Resources)
    // Formerly /api/librarybook vs /api/library-book. Fixed to /api/library-book.
    // Service calls GET /library-book (proxy -> /api/library-book)
    console.log('\nTesting Library Books (/api/library-book)...');
    try {
        const res = await fetch(`${baseUrl}/api/library-book`, {
            method: 'GET',
            headers: headers
        });
        if (res.ok) console.log(`[PASS] Library Books (GET): ${res.status}`);
        else {
            // Try fallback search?
            console.log(`[INFO] GET failed (${res.status}), trying SEARCH...`);
            const searchRes = await fetch(`${baseUrl}/api/library-book/search`, {
                method: 'POST',
                headers: headers,
                body: JSON.stringify({})
            });
            if (searchRes.ok) console.log(`[PASS] Library Books (SEARCH): ${searchRes.status}`);
            else console.error(`[FAIL] Library Books All Failed: ${searchRes.status} ${await searchRes.text()}`);
        }
    } catch (e) { console.error(`[FAIL] Library Books Error: ${e.message}`); }

    // 4. Test Inventory (Resources)
    // /api/inventory-item
    console.log('\nTesting Inventory (/api/inventory-item)...');
    try {
        const res = await fetch(`${baseUrl}/api/inventory-item`, {
            method: 'GET',
            headers: headers
        });
        if (res.ok) console.log(`[PASS] Inventory (GET): ${res.status}`);
        else console.error(`[FAIL] Inventory: ${res.status}`);
    } catch (e) { console.error(`[FAIL] Inventory Error: ${e.message}`); }

    // 5. Test Transport (Get Available)
    console.log('\nTesting Transport Available (/api/transport/available)...');
    try {
        const res = await fetch(`${baseUrl}/api/transport/available`, {
            method: 'GET',
            headers: headers
        });
        if (res.ok) console.log(`[PASS] Transport Available: ${res.status}`);
        else console.error(`[FAIL] Transport Available: ${res.status}`);
    } catch (e) { console.error(`[FAIL] Transport Error: ${e.message}`); }

    // 6. Test Academic (Course Search)
    console.log('\nTesting Course Search (/api/course/search)...');
    try {
        const res = await fetch(`${baseUrl}/api/course/search`, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify({})
        });
        if (res.ok) console.log(`[PASS] Course Search: ${res.status}`);
        else console.error(`[FAIL] Course Search: ${res.status}`);
    } catch (e) { console.error(`[FAIL] Course Search Error: ${e.message}`); }
}

async function run() {
    await testEndpoints('http://localhost:3003');
}

run();
